### PR TITLE
[WIP] Add disposition to download elements

### PR DESCRIPTION
**Pull request by Defcon0**
PR for ticket

https://github.com/contao/contao/issues/14

(Now in the correct branch :-) )

### CODE REVIEW

**Comment by fritzmg** [comment]
May be the description of your select and the options should be more user friendly. Regular users might not know the difference.

**Comment by Defcon0** [comment]
OK, you might be right ;) What about:

Field title:
```
de: Auslieferung
en: Disposition
```

Field hint:
```
de: Wählen Sie hier aus, ob der Download nach einem Klick als Datei ausgespielt, oder direkt im Browser ausgegeben werden soll.
en: Choose whether the download should be disposed as a downloadable file or directly in the browser.
```

Options:
```
attachment: de: Als Download ausspielen, en: Dispose as download
inline: de: Im Browser öffnen, en: Dispose in browser
```

**Comment by fritzmg** [comment]
Yeah, that's better ;)

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 1]
```diff
@@ -125,8 +125,8 @@
...
 		'youtube'                     => '{type_legend},type,headline;{source_legend},youtube;{player_legend},playerSize,youtubeOptions,playerStart,playerStop,playerCaption,playerAspect;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID;{invisible_legend:hide},invisible,start,stop',
 		'vimeo'                       => '{type_legend},type,headline;{source_legend},vimeo;{player_legend},playerSize,vimeoOptions,playerStart,playerColor,playerCaption,playerAspect;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID;{invisible_legend:hide},invisible,start,stop',
-		'download'                    => '{type_legend},type,headline;{source_legend},singleSRC,overwriteLink;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID;{invisible_legend:hide},invisible,start,stop',
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_content.php]: Why did you remove this?

**Comment by Defcon0** [comment]
Sorry, because I copied it from 4.4 ;-) I‘ll fix it tomorrow in addition to the language fixes.

**Comment by Defcon0** [comment]
OK, I added a new commit containing the fixed.

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 2]
```diff
@@ -501,6 +501,18 @@
...
+			'label'                   => &$GLOBALS['TL_LANG']['tl_content']['disposition'],
+			'exclude'                 => true,
+			'inputType'               => 'select',
+			'options' => [
+				\Symfony\Component\HttpFoundation\ResponseHeaderBag::DISPOSITION_ATTACHMENT,
+				\Symfony\Component\HttpFoundation\ResponseHeaderBag::DISPOSITION_INLINE
+			],
+			'reference'               => &$GLOBALS['TL_LANG']['MSC']['disposition'],
+			'eval'                    => ['tl_class' => 'w50'],
+			'sql'                     => "varchar(32) NOT NULL default '" . \Symfony\Component\HttpFoundation\ResponseHeaderBag::DISPOSITION_ATTACHMENT . "'"
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/dca/tl_content.php]: @leofeyer what is best practise here within Contao? use
```php
'sql' => "varchar(32) NOT NULL default ''"
```
together with
```php
'default' => ResponseHeaderBag::DISPOSITION_ATTACHMENT
```
? Usually the defaults within Contao are an empty string or null. Even if the field stays empty, the default will still be `attachment` due to the backwards compatible implementation within `sendFileToBrowser`.

Also should one use the FQCN within the DCA or use
```php
use Symfony\Component\HttpFoundation\ResponseHeaderBag;
```
?

**Comment by asaage** [comment]
just for the record:
https://github.com/contao/core/issues/2382

**Comment by Defcon0** [comment]
If this is merged, will it also find its way into the 4.4 branch since it's a new feature? I fear, it won't, will it?

**Comment by Toflar** [comment]
No way, it's a new feature so it'll be 4.7

**Comment by Defcon0** [comment]
OK, clear. No problem ;-)

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 3]
```diff
@@ -125,8 +125,8 @@
...
-		'download'                    => '{type_legend},type,headline;{source_legend},singleSRC,overwriteLink;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID;{invisible_legend:hide},invisible,start,stop',
-		'downloads'                   => '{type_legend},type,headline;{source_legend},multiSRC,sortBy,metaIgnore;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID,useHomeDir;{invisible_legend:hide},invisible,start,stop',
+		'download'                    => '{type_legend},type,headline;{source_legend},singleSRC,overwriteLink;{dwnconfig_legend},disposition;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID;{invisible_legend:hide},invisible,start,stop',
+		'downloads'                   => '{type_legend},type,headline;{source_legend},multiSRC,sortBy,metaIgnore;{dwnconfig_legend},disposition;{template_legend:hide},customTpl;{protected_legend:hide},protected;{expert_legend:hide},guests,cssID,useHomeDir;{invisible_legend:hide},invisible,start,stop',
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/dca/tl_content.php]: This legend looks awkward to me, we never use such strange abbreviations. why not just `download_legend`?

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 2 above)
Comment (Toflar) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: FQCN is good here. It's a config array, not a class.

File: core-bundle/src/Resources/contao/languages/de/default.xlf [hunk 4]
```diff
@@ -1957,6 +1957,14 @@
         <source>descending</source>
         <target>absteigend</target>
       </trans-unit>
+      <trans-unit id="MSC.disposition.attachment">
+        <source>Dispose as download</source>
+        <target>Als Download ausspielen</target>
+      </trans-unit>
+      <trans-unit id="MSC.disposition.inline">
+        <source>Dispose in browser</source>
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/languages/de/default.xlf]: I don't like`dispose` here. To me it sounds like throwing something away. I'd prefer something like `Provide as attachment (download)` and `Provide inline (within Browser)`.

File: core-bundle/src/Resources/contao/languages/de/tl_content.xlf [hunk 5]
```diff
@@ -297,6 +297,14 @@
         <source>The link title is added as &lt;em&gt;title&lt;/em&gt; attribute in the HTML markup.</source>
         <target>Der Link-Titel wird als &lt;em&gt;title&lt;/em&gt;-Attribut im HTML-Markup eingefügt.</target>
       </trans-unit>
+      <trans-unit id="tl_content.disposition.0">
+        <source>Disposition</source>
+        <target>Auslieferung</target>
+      </trans-unit>
+      <trans-unit id="tl_content.disposition.1">
+        <source>Choose whether the download should be disposed as a downloadable file or directly in the browser.</source>
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/languages/de/tl_content.xlf]: `disposed` -> `provided`

File: core-bundle/src/Resources/contao/elements/ContentDownload.php [hunk 6]
```diff
@@ -66,7 +66,7 @@ public function generate()
 		{
 			if ($file == $objFile->path)
 			{
-				\Controller::sendFileToBrowser($file);
+				\Controller::sendFileToBrowser($file, $this->disposition);
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/elements/ContentDownload.php]: You will need a fallback here for already existing elements in which case `$this->disposition` will be `null`.

File: core-bundle/src/Resources/contao/elements/ContentDownloads.php [hunk 7]
```diff
@@ -77,7 +77,7 @@ public function generate()
 			{
 				if ($file == $this->objFiles->path || \dirname($file) == $this->objFiles->path)
 				{
-					\Controller::sendFileToBrowser($file);
+					\Controller::sendFileToBrowser($file, $this->disposition);
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/elements/ContentDownloads.php]: You will need a fallback here for already existing elements in which case `$this->disposition` will be `null`.

File: core-bundle/src/Resources/contao/languages/de/tl_content.xlf [hunk 8]
```diff
@@ -931,4 +943,4 @@
       </trans-unit>
     </body>
   </file>
-</xliff>
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/languages/de/tl_content.xlf]: We always have new lines at the end of every file. You have to configure this in your IDE settings.

**Review by Toflar — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 2 above)
Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: Sure, but is the usage of "use" limited to classes? I always use "use" in DCAs as well to keep them concise.

Comment (Toflar) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: I find it way harder to debug. That's why I prefer FQCN in DCA's. Matter of preference, there's no rule and I don't care really.

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 3 above)
Comment (Defcon0) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: I just took the old name from contao 4.4.21 ;-)

File: core-bundle/src/Resources/contao/elements/ContentDownload.php (hunk 6 above)
Comment (Defcon0) [reply core-bundle/src/Resources/contao/elements/ContentDownload.php]: Existing elements get attachment as the db value, so this won't be necessary, no? Also in Controller::sendFileToBrowser I built in a default value for the parameter (`$strDisposition = ResponseHeaderBag::DISPOSITION_ATTACHMENT`).

File: core-bundle/src/Resources/contao/elements/ContentDownloads.php (hunk 7 above)
Comment (Defcon0) [reply core-bundle/src/Resources/contao/elements/ContentDownloads.php]: Same here, see above.

File: core-bundle/src/Resources/contao/languages/de/tl_content.xlf (hunk 8 above)
Comment (Defcon0) [reply core-bundle/src/Resources/contao/languages/de/tl_content.xlf]: OK :)

File: core-bundle/src/Resources/contao/elements/ContentDownload.php (hunk 6 above)
Comment (fritzmg) [reply core-bundle/src/Resources/contao/elements/ContentDownload.php]: The default would not be in use, if you pass `null` as the second parameter. May be you should make `null` the default and then check for `null` for that parameter and use `ResponseHeaderBag::DISPOSITION_ATTACHMENT` as the default.

Comment (Defcon0) [reply core-bundle/src/Resources/contao/elements/ContentDownload.php]: Right, but as mentioned before, we have the default value in the sql dca key. Would be enough, no? Passing in null would only be possible if one used the function statically from his own code. And then he would know the function's signature.

**Comment by Defcon0** [comment]
OK, I fixed the desired lines. Only the fallback for `Controller::sendFileToBrowser()` is left for discussion.

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: Empty default is the way to go, otherwise every row will contain `attachment`, regardless of whether an element is a download element or not.

File: core-bundle/src/Resources/contao/elements/ContentDownload.php (hunk 6 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/elements/ContentDownload.php]: It is not enough. You need to either make `null` the default and handle it in the method body or you need to pass `$this->disposition ?? ResponseHeaderBag::DISPOSITION_ATTACHMENT`.

**Review by leofeyer — CHANGES_REQUESTED** [review]
Please only change the English translation files. All other languages are maintained on Transifex.

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: So? I don't see why that would be a problem? Every field has a gallery list of 4 items by default, regardless of whether it's a gallery element or not.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: That's a bad example, because storing `4` instead of `0` does not affect the database size. But here are some fields we do need to fix:

 * `tl_content.sliderSpeed` stores `300` in every row (should be `0`)
 * `tl_module.queryType` stores `and` in every row (should be `''`)
 * `tl_module.searchType` stores `simple` in every row (should be `''`)
 * `tl_page.sitemap` stores `map_default` in every row (should be `''`) 

However, we should open a separate ticket hereto.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: See contao/contao#56

**Comment by Defcon0** [comment]
OK, I fixed the fallback and the database value for the disposition field in tl_content. But how/where do I fix the transifex issues?

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: As discussed in Mumble on September 27th, we are currently not moving the defaults to the SQL definition, therefore the default needs to go into the `'default'` value.

**Comment by Defcon0** [comment]
As stated in https://github.com/contao/contao/pull/20#issuecomment-421934973 I completed all desired fixes. Only the transifex issue is missing. Where do I have to put my changes?

**Comment by ausi** [comment]
> Only the transifex issue is missing. Where do I have to put my changes?

Please revert the changes in *languages/de/default.xlf* and *languages/de/tl_content.xlf* and instead add the English translations to *languages/en/default.xlf* and *languages/en/tl_content.xlf*.

The German translations will be done on Transifex afterwards.

**Comment by Defcon0** [comment]
@ausi OK, done.

**Comment by leofeyer** [comment]
Thank you @Defcon0.

**Comment by zonky2** [comment]
@Defcon0 

Hi, I saw the new feature in the news - great!

Addendum: wouldn't it be great if the checkbox "show in browser" was clicked to add a new checkbox "open in new window/tab" for `target="_blank"`?

**Comment by Defcon0** [comment]
Sounds reasonable to me :-)
